src/bilder/components/docker/steps.py

```diff
@@ -0,0 +1,78 @@
+import json
+from io import StringIO
+
+from pyinfra import host
+from pyinfra.api.deploy import deploy
+from pyinfra.api.exceptions import DeployError
+from pyinfra.api.util import make_hash
+from pyinfra.facts.deb import DebPackages
+from pyinfra.facts.server import Command, LinuxDistribution
+from pyinfra.operations import apt, files
+
+
+def _apt_install():
+
+    apt.packages(
+        name="Ensure Docker CE prerequisites are present",
+        present=True,
+        packages=["curl", "gnupg-agent", "software-properties-common", "lsb-release"],
+        update=True,
+    )
+
+    apt.packages(
+        name="Install apt requirements to use HTTPS",
+        present=True,
+        packages=["apt-transport-https", "ca-certificates"],
+        update=True,
+        cache_time=3600,
+    )
+    lsb_release = host.get_fact(LinuxDistribution)
+    lsb_id = lsb_release["name"].lower()
+
+    apt.key(
+        name="Download the Docker apt key",
+        src=f"https://download.docker.com/linux/{lsb_id}/gpg",
+    )
+
+    dpkg_arch = host.get_fact(Command, command="dpkg --print-architecture")
+
+    add_apt_repo = apt.repo(
+        name="Add the Docker apt repo",
+        src=(
+            f"deb [arch={dpkg_arch}] https://download.docker.com/linux/{lsb_id}"
+            f" {lsb_release['release_meta']['CODENAME']} stable"
+        ),
+        filename="docker-ce-stable",
+    )
+
+    apt.packages(
+        name="Install Docker via apt",
+        packages="docker-ce",
+        update=add_apt_repo.changed,
+    )
+
+
+@deploy("Deploy Docker")
+def deploy_docker(config=None):
```

Comment:
We generally try to take advantage of type hints so it's probably worth adding `config: Optional[Dict[str, Any]] = None` rather than using the `isinstance` check further down.